### DISCUSSION TITLE
tutorials/yolo-class-balancing/

### DISCUSSION THREAD

**Discussion by giscus** [Y-T-G/Yasins-Keep#7] [Giscus]
# tutorials/yolo-class-balancing/

Implement class balancing in Ultralytics using a weighted dataloader and improve the performance of minority class.

https://y-t-g.github.io/tutorials/yolo-class-balancing/

**Comment by Jordan-Pierce** [comment]
If anyone needs this same functionality but for Classification:

```python

from ultralytics.data.dataset import ClassificationDataset
import ultralytics.models.yolo.classify.train as build

class WeightedClassificationDataset(ClassificationDataset):
    def __init__(self, *args, mode='train', **kwargs):
        
"""
        Initialize the WeightedClassificationDataset.
        Args:
            class_weights (list or numpy array): A list or array of weights corresponding to each class.
        """
        
super(WeightedClassificationDataset, self).__init__(*args, **kwargs)

        self.train_mode = "train" in self.prefix

        self.count_instances()
        class_weights = np.sum(self.counts) / self.counts

        # Aggregation function
        self.agg_func = np.mean

        self.class_weights = np.array(class_weights)
        self.weights = self.calculate_weights()
        self.probabilities = self.calculate_probabilities()

    def count_instances(self):
        
"""
        Count the number of instances per class
        Returns:
            dict: A dict containing the counts for each class.
        """
        
self.counts = [0 for i in range(len(self.base.classes))]
        for _, class_idx, _, _ in self.samples:
            self.counts[class_idx] += 1
        self.counts = np.array(self.counts)
        self.counts = np.where(self.counts == 0, 1, self.counts)

    def calculate_weights(self):
        
"""
        Calculate the aggregated weight for each label based on class weights.
        Returns:
            list: A list of aggregated weights corresponding to each label.
        """
        
weights = []
        for _, class_idx, _, _ in self.samples:
            weight = self.agg_func(self.class_weights[class_idx])
            weights.append(weight)
        return weights

    def calculate_probabilities(self):
        
"""
        Calculate and store the sampling probabilities based on the weights.
        Returns:
            list: A list of sampling probabilities corresponding to each label.
        """
        
total_weight = sum(self.weights)
        probabilities = [w / total_weight for w in self.weights]
        return probabilities

    def __getitem__(self, index):
        
"""
        Return transformed label information based on the sampled index.
        """
        
if self.train_mode:
            index = np.random.choice(len(self.samples), p=self.probabilities)

        return super(WeightedClassificationDataset, self).__getitem__(index)

build.ClassificationDataset = WeightedClassificationDataset
target_model = YOLO(model_path)
target_model.train(**params)

```

**Comment by ilyasofficial1617** [comment]
How do you use this? Do we use it to train as usual?
Does it affect all the dataloaders in both the library and the CLI?

**Reply by Y-T-G** [reply]
I have linked the Colab notebook in the beginning with the training process shown. You run it using Python. It wouldn't work with CLI because you have to run the monkey-patching code first and you have to do it every time.

**Comment by mdminhazurrahman** [comment]
NameError: name 'YOLODataset' is not defined when I use monkey patch. Do you have any suggestion? thank you!

**Reply by Y-T-G** [reply]
Did you follow the Colab notebook?

Are you using multi-GPU training?

**Reply by mdminhazurrahman** [reply]
now it's working watching the colab nb. Yes I am using multi gpu. Wont it work for gpus?

**Reply by mdminhazurrahman** [reply]
Could you please tell where should I put this line for multigpu?
build.YOLODataset = YOLOWeightedDataset

**Reply by Y-T-G** [reply]
I have the training instruction for multi-GPU at the end of the Colab notebook.

**Comment by Areebkhan02** [comment]
Hello Yasin, thanks for this amazing blog post. I have a peculiar question that is not related to this blog post but in general. I have trained my Yolov8 model on my brand logos like Nike, Adidas and others. One thing I have noticed is that the model detects the distinguished features of the brand like Adidas symbol or Nike tick sign but when written plainly Nike or Adidas it faces difficulty even thou I have trained the model on plain text nike and adidas as well. SImilar things happen to other plain text as well so is there any particular reason for it ?

**Reply by Y-T-G** [reply]
I'm not sure. It should work if there's enough data. Did you disable the flipping augmentations?

You can join the Ultralytics discord if you want to ask anything related to Ultralytics models.

**Reply by Areebkhan02** [reply]
Are there any reditt or linkedIn groups for ultralytics questions ? or General CV questions where community can collaborate and learn 

One the side note nope I didn't tuned it off infact I did more augmentation which involves flipping, rotating to increase more data will it have any effect on the accuracy or in particular detections

**Reply by Y-T-G** [reply]
There's Ultralytics [subreddit](https://reddit.com/r/ultralytics). And you can ask questions like that on the [Discord](https://discord.com/invite/ultralytics) server too. They also have a [forum](https://community.ultralytics.com).

**Comment by pauliwag** [comment]
Hi Yasin. When I run the following code after training (i.e., after `model.train()`) to visualize class distribution in normal vs. weighted modes: 
```
model.trainer.train_loader.dataset.agg_func = (
    np.mean
) # Can be np.max, np.mean, np.median, or np.sum
model.trainer.train_loader.dataset.weights = (
    model.trainer.train_loader.dataset.calculate_weights()
)
model.trainer.train_loader.dataset.probabilities = (
    model.trainer.train_loader.dataset.calculate_probabilities()
)

# Verify class balance in training mode
model.trainer.train_loader.dataset.train_mode = True
train_counts = verify_class_balance(
    model.trainer.train_loader.dataset, num_samples=1000
)

# Verify class balance in validation mode
model.trainer.train_loader.dataset.train_mode = False
val_counts = verify_class_balance(
    model.trainer.train_loader.dataset, num_samples=1000
)

# Plot the comparison
plot_class_balance(
    train_counts,
    val_counts,
    set(model.trainer.train_loader.dataset.data["names"].values()),
)
```
I get this error: 
```
Traceback (most recent call last):
  File "train.py", line 262, in <module>
    results = main()
  File "train.py", line 224, in main
    model.trainer.train_loader.dataset.agg_func = (
AttributeError: 'DetectionTrainer' object has no attribute 'train_loader'
```
Since you wrote this, have they made `train_loader` no longer directly accessible or something?  Do you know how this should be updated?  Thanks so much for any help you're willing to give on this.

**Reply by Y-T-G** [reply]
It still works for me. You can try starting the training again and cancelling it once it starts the epoch. And then running the code for visualization again. It would still show the balanced counts even if the training wasn't completed.

**Reply by pauliwag** [reply]
Thanks for such a quick reply!  It seems to be an environment issue – I was originally trying to run a training script in Conda on Linux, but I migrated the code to Colab and now it works. 

Btw, for a large dataset like VisDrone (6471 train images, 548 val, 1610 test), should `range(len(dataset.labels))` in the `verify_class_balance()` conditional be changed to `np.random.choice(len(dataset.labels), size=num_samples)`?  
And `val_counts` (for verifying balance in validation/nonweighted mode) also samples from the train dataset, correct (so for VisDrone I could keep `num_samples=1000` for both `train_counts` and `val_counts`)?

And one last question, I hope you won't mind 😅  If I wish to try median aggregation instead of mean, do I only change
- `self.agg_func = np.sum` to `self.agg_func = np.median` in `__init__()`, and 
- `model.trainer.train_loader.dataset.agg_func = np.mean` to `model.trainer.train_loader.dataset.agg_func = np.median` in the balance verification code, 

or do I also need to change any other parts of the code?  (When I just change `np.mean` to `np.median` in the balance verification code, the weighted bars appear slightly more balanced for VisDrone.)

Thank you so much, truly, for any insights you might be willing to give on these.

**Reply by Y-T-G** [reply]
> Thanks for such a quick reply! It seems to be an environment issue – I was originally trying to run a training script in Conda on Linux, but I migrated the code to Colab and now it works.
> 
> Btw, for a large dataset like VisDrone (6471 train images, 548 val, 1610 test), should `range(len(dataset.labels))` in the `verify_class_balance()` conditional be changed to `np.random.choice(len(dataset.labels), size=num_samples)`? And `val_counts` (for verifying balance in validation/nonweighted mode) also samples from the train dataset, correct (so for VisDrone I could keep `num_samples=1000` for both `train_counts` and `val_counts`)?
> 

Yes, changing to `np.random.choice(len(dataset.labels), size=num_samples)` should be better and yes both uses training dataset.

> And one last question, I hope you won't mind 😅 If I wish to try median aggregation instead of mean, do I only change
> 
> * `self.agg_func = np.sum` to `self.agg_func = np.median` in `__init__()`, and
> * `model.trainer.train_loader.dataset.agg_func = np.mean` to `model.trainer.train_loader.dataset.agg_func = np.median` in the balance verification code,
> 
> or do I also need to change any other parts of the code? (When I just change `np.mean` to `np.median` in the balance verification code, the weighted bars appear slightly more balanced for VisDrone.)

Yes, you just change `self.agg_func` to a different one.

**Reply by Y-T-G** [reply]
@pauliwag I updated the code. So it uses `np.random.choice(len(dataset.labels), size=num_samples, replace=False)` now when not in `train_mode`. We also need to disable replacement since that doesn't occur in default mode.

**Comment by Dh75-2004** [comment]
```
New https://pypi.org/project/ultralytics/8.3.11 available  Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.2.93  Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 49140MiB)
engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=data.yaml, epochs=50, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=0, workers=8, project=None, name=train2, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs\detect\train2
Overriding model.yaml nc=80 with nc=2

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]           
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]                  
 16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]                 
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 21                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]                 
 22        [15, 18, 21]  1    751702  ultralytics.nn.modules.head.Detect           [2, [64, 128, 256]]           
Model summary: 225 layers, 3,011,238 parameters, 3,011,222 gradients, 8.2 GFLOPs

Transferred 319/355 items from pretrained weights
TensorBoard: Start with 'tensorboard --logdir runs\detect\train2', view at http://localhost:6006/
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed 
train: Scanning D:\final\new Aug.v1i.yolov8\train\labels.cache... 2649 images, 0 backgrounds, 0 corrupt: 100%|█████████
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, num_output_channels=3, method='weighted_average'), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))
```




"WE TRIED THE SAME CODE AS YOU PROVIDED. BUT WHEN RUN THE CODE, THEN IT IS TAKING SO MUCH TIME AND GETTING STUCKED AT THIS POINT OF ALBUMENTATION. WE ARE TRYING TO RUN THE CODE ON OUR SYSTEM (JUPYTER NOTEBOOK)."

**Reply by Y-T-G** [reply]
Stop and restart the training.

**Reply by AnkitSharma1405** [reply]
Hello yasin !

I am also working with Dh75-2004, Like we have tried to run it so many times.It is getting stucked always. All our models are working fine on our system. But when we try to run the 
code for weighted dataloader ( as provided by you), We are facing problem.

**Reply by Y-T-G** [reply]
You can add some print statements in `YOLOWeightedDataset` `__init__` method and find out at which point it gets stuck.

**Comment by Mahim-SOL** [comment]
Hello Yasin, I am new in YOLO and so pardon If I am asking something so trivial. My question is if I want to use this weighted dataloader for a custom dataset (for my case xBD dataset), how can I do it?

**Reply by Y-T-G** [reply]
You can follow the Colab notebook and just change the dataset to the one you're using.

**Comment by hyunjun97** [comment]
Is this method valid for segmentation tasks?
When I ran the provided code, I encountered the following error:

```
AttributeError                            Traceback (most recent call last)
Cell In[13], line 2
      1 # Verify the type of dataset used
----> 2 model.trainer.train_loader.dataset
AttributeError: 'SegmentationTrainer' object has no attribute 'train_loader'
```

**Reply by Y-T-G** [reply]
Did you run that code after training? It should work.

**Comment by Jordan-Pierce** [comment]
Dump question, but does the model file (.pt) somehow store the data loader used during training?

**Reply by Y-T-G** [reply]
No

**Comment by mings-k** [comment]
Thank you for your nice explain but can you tell me how to use this code to rt-detr??

**Reply by Y-T-G** [reply]
You can check [this](https://github.com/Y-T-G/Yasins-Keep/discussions/7#discussioncomment-11593743).

**Comment by azmiozgen** [comment]
Thank you for this great post. I wonder about something in the `__getitem__` function. By modifying the `index` value, how do we ensure that the majority class samples are not taken more than once? I know that minority samples must be taken more than once across the batches. But shouldn't we avoid choosing the same samples and ensure that at least all of the samples of a class are taken at least once in one pass of the dataloader? Or is there something that I'm missing that prevents these repetitions?

**Reply by Y-T-G** [reply]
It doesn't ensure all the samples are taken at least once in this implementation. The length of the dataloader is unchanged, i.e., it produces the same number of images as the normal dataloader every epoch. So some samples would have to be dropped every epoch in order to accommodate minority samples. You can of course increase the length of the dataloader and make it select all the samples at least once, but why not just run more epochs in that case? Since you're simply trading more epochs for longer epochs.

**Comment by abdelazizashraf123** [comment]
how to use this code to rt-detr ???

**Reply by Y-T-G** [reply]
You can load `RTDETRDataset` instead of `YOLODataset`.

And then use it as parent class:
```python
from ultralytics.models.rtdetr.val import RTDETRDataset

class YOLOWeightedDataset(RTDETRDataset):
# Same code as in the guide
```

Then monkey-patch using:

```python
import ultralytics.models.rtdetr.train as train

train.RTDETRDataset = YOLOWeightedDataset
```

Then train normally,

**Reply by tejasri19** [reply]
What about yolov6 and yolov7 as they have different method of dataloader?

**Reply by Y-T-G** [reply]
If you use Ultralytics, it's going to be the same as what's in the guide.

**Reply by tejasri19** [reply]
For yolov6 & v7, the folders are to be cloned and run in CLI as follows:
python train.py --workers 8 --device 0 --batch-size 32 --data data/custom.yaml --img 640 640 --cfg cfg/training/yolov7-custom.yaml --weights 'yolov7_training.pt' --name yolov7-custom --hyp data/hyp.scratch.custom.yaml

In this case, how to do monkey patch the dataloader?

**Reply by Y-T-G** [reply]
What I showed was for Ultralytics only. YOLOv6 or YOLOv7 repo is not Ultralytics.

**Reply by tejasri19** [reply]
Any directions on how to do the same for v6 & v7 variants?

**Reply by Y-T-G** [reply]
You can use Ultralytics for YOLOv6.

**Reply by tejasri19** [reply]
There is no direct package for yolov6 by ultralytics. Have to clone respective repository!

**Reply by Y-T-G** [reply]
You can use YOLOv6 with the Ultralytics library.

https://docs.ultralytics.com/models/yolov6/#usage-examples

[a comment by an automated account is not shown]

**Reply by Y-T-G** [reply]
I have the visualization in the Colab notebook at the end.

**Reply by Y-T-G** [reply]
`num_samples` controls how many of the samples it takes for comparison. If you use 1000, it will take the first 1000 images from the default training dataloader and the first 1000 images from weighted dataloader, and compare their class distribution.

**Comment by speendo** [comment]
Hello and thank you for this interesting approach!

I try for one week to get it running on a multi-gpu environment **with yolov11**. I used callbacks just as you suggest in the Colab. However, I still cannot get the balanced data loader to be used in my training.

Can you give me some insight on this?

Btw. I also discussed the issue here:
https://community.ultralytics.com/t/balance-classes-during-training-yolov11/692

Thank you in advance for your help!

**Reply by Y-T-G** [reply]
I updated the multi-GPU workaround to modify source code. You can test the new code in Colab.

**Reply by speendo** [reply]
Thanks. I got it working by orverwriting the original library

**Reply by VG-Fish** [reply]
Do you know how can I write a custom data prefetcher for YOLO models?

**Reply by Y-T-G** [reply]
You will have to check the Ultralytics source code. Ultralytics uses a highly customized dataloader. So it's not straightforward to write a custom one that's compatible with Ultralytics.

**Reply by VG-Fish** [reply]
I would mainly like to experiment with data prefetching. I'll checkout the source code.

[a comment by an automated account is not shown]

**Comment by isDavidd976** [comment]
Hello Yasin, this is a great tutorial! Are you planning to write a short technical report or paper on this, similar to your other tutorial on [Extending YOLO Models With New Classes Through Additional Head](http://dx.doi.org/10.13140/RG.2.2.12569.53609)? I am doing some research, which might use your method to mitigate class imbalanced problem. So if you're planning to write a short technical report on ResearchGate, then it will be easier for me (and other researchers) to citate your work. Else, I will just citate this webpage. Anyway, keep up the good work!

**Reply by Y-T-G** [reply]
Hello David. Thanks for reading.

I am not planning to write a technical report on this. You can use the citation at the bottom of the page if needed.

**Comment by jksim1833** [comment]
Thanks for great work! 
I'm trying to use multi gpu case, but always get same problem. 


ImportError: cannot import name 'ARM64' from partially initialized module 'ultralytics.utils' (most likely due to a circular import) (/home/work/yolo_custom/ultralytics/ultralytics/utils/__init__.py)

I also tried on the colab (only for the gpus part). I got a same error message
```
ImportError                               Traceback (most recent call last)
<ipython-input-8-dd1592353afd> in <cell line: 0>()
----> 1 from ultralytics import YOLO
      2 from ultralytics.data.build import YOLODataset

14 frames
/content/ultralytics/ultralytics/nn/autobackend.py in <module>
     15 from PIL import Image
     16 
---> 17 from ultralytics.utils import ARM64, IS_JETSON, LINUX, LOGGER, PYTHON_VERSION, ROOT, YAML
     18 from ultralytics.utils.checks import check_requirements, check_suffix, check_version, check_yaml, is_rockchip
     19 from ultralytics.utils.downloads import attempt_download_asset, is_url

ImportError: cannot import name 'ARM64' from partially initialized module 'ultralytics.utils' (most likely due to a circular import) (/content/ultralytics/ultralytics/utils/__init__.py)

```

Do you have any idea about it?

**Reply by Y-T-G** [reply]
Did you try the code at the bottom of the Colab notebook for multi-GPU?

**Reply by jksim1833** [reply]
I tried both on my server (gpus) and colab notebook for multi-GPU, but got the same error message about ultralytics.utils
I even tried to allocate the python path like below or downgrade the ultralytics version as like yours and edited on the library side (not from the clonning the repository)
[import sys
sys.path.insert(0, '/path/to/local/ultralytics') ]
I'm just wondering if you can check the colab notebook multi-gpu code, cause I got the same error message from my local server

**Reply by Y-T-G** [reply]
You can try the updated notebook.

**Reply by jksim1833** [reply]
Thanks a lot bro

**Comment by UnlivedDisc03** [comment]
My yolov8 wandb logs for my class distribution (same as your "Class distribution of the traffic lights dataset") after monkey patching the weighted datasert look the same as before. Is the weighted dataset being correctly implemented? Everything works as it should, all confirmations that its using the weighted dataset are there but my minimal difference in result as well as the graph on wandb logs got me questioning it.

**Reply by Y-T-G** [reply]
Which logs are you referring to? The `labels.jpg` file? That wouldn't be affected by the change.

**Comment by sparking-cloud** [comment]
Can you tell me how to port the code you provided into yolov8? And, how to modify the weight part?

**Reply by Y-T-G** [reply]
There's no difference. This code would work for all YOLO models in Ultralytics.

**Reply by sparking-cloud** [reply]
What I want to know is if I just need to modify the YOLODataset in ultralytics/data/build.py to change it to YOLOWeightedDataset, and no need to modify the other parts?

**Reply by Y-T-G** [reply]
I have a notebook link provided in the guide. You just follow the steps. There's no modification needed.

**Comment by TOF4T** [comment]
Hi Yasin,
Thanks for this amazing post. I have a question about algorithms. Is the algorithm you are using influenced by Repeat Factor Sampling, or by another algorithm?

**Reply by Y-T-G** [reply]
I didn't look at other algorithms for this. The one I used here is just what came to my mind and made sense. I would be interested to know if you are able to get better results with any other algorithm.

**Reply by TOF4T** [reply]
I used model yolov11l and seen it not better than other algorithm I used. But when I combined yours, Multilabel Stratification and augmentation less class, the performance is very good with mean F1-Score is 0.9275 and mAP50 is 0.9301.

**Reply by TOF4T** [reply]
I'm sorry because my English not good enough to explain to you.

**Reply by Y-T-G** [reply]
Oh cool. What was the mAP before applying any class balancing?

Ultralytics also recently added `cls_pw` argument which applies class weights to the BCE loss based on inverse frequency. You can also try that.

**Reply by TOF4T** [reply]
mAP before apply balance is 0.8227.
Thanks for the info, I will test that.
